README.md
# ledger-thorchain-js

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

This package provides a basic client library to communicate with a THORChain App running in a Ledger Nano S/X

We recommend using the [npm package](https://www.npmjs.com/package/@thorchain/ledger-thorchain) in order to receive updates/fixes.

## Build

```bash
yarn build
```

## Tests

### Unit tests

```bash
yarn test:unit
```

### e2e tests

`e2e' tests are to run some tests against your connected Ledger locally (no CI).

Create a `.e2e.env` (use `.e2e.env.sample` as template) to add `THOR_PUB_KEY` (public key of your Ledger) and `THOR_ADDRESS` (`thor` address of your Ledger).

Connect your Ledger, open `THORChain` app on it and start e2e tests locally:

```bash
yarn test:e2e
```

Hint: Add [`.only`](https://jestjs.io/docs/api#testonlyname-fn-timeout) to the tests to run a single test only


## Example

This repo also includes a simple Vue example for WebUSB and WebHID.

To run it locally

```bash
yarn serve
```
and open Chrome at https://localhost:8080/

![Example](docs/example.png)


CHANGELOG.md
# 0.1.0 (2021-XX-XX)

## Update

- Rewrite in TypeScript
- Introduce custom `LedgerError`
- Improve `Promise` handling

## Breaking changes

- Values of misc. `XYZResponse`'s have been renamed to use camel case (instead of snake case)
- Re-organize helper function to be available in `helpers` and `common` only (Deprecated `v1` helpers have been removed)
- Rename some methods of THORChain app (mostly to prefix getter methods with `get...`)

# 0.0.3 (2021-07-20)

## Internal updates

- Update dependencies, Remove deprecated `@ledgerhq/hw-transport-u2f` from examples, Use same `jest.setTimeout` for all `e2e` tests [#2](https://github.com/thorchain/ledger-thorchain-js/pull/2)
- Update tests [#1](https://github.com/thorchain/ledger-thorchain-js/pull/1)

# 0.0.2 (2021-06-30)

Initial release


tests/e2e/basic.spec.ts
import THORChainApp, { LedgerErrorType } from "../../src/index";
import TransportNodeHid from "@ledgerhq/hw-transport-node-hid";
import secp256k1 from "secp256k1";
import crypto from "crypto";
import Transport from "@ledgerhq/hw-transport";

require("dotenv").config({ path: "./.e2e.env" });

const { THOR_ADDRESS = "unknown address", THOR_PUB_KEY = "unknown pub key" } = process.env;

const JEST_TIMEOUT = 60000;
// Use same time out for all tests in this file
jest.setTimeout(JEST_TIMEOUT);

describe("THORChainApp e2e", () => {
  let transport: Transport;
  let app: THORChainApp;
  beforeAll(async () => {
    transport = await TransportNodeHid.create(1000);
    app = new THORChainApp(transport);
  });

  afterAll(async () => {});

  it("getVersion", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****
    const resp = await app.getVersion();

    console.log(resp);

    expect(resp.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(resp.errorMessage).toEqual("No errors");
    expect(resp).toHaveProperty("testMode");
    expect(resp).toHaveProperty("major");
    expect(resp).toHaveProperty("minor");
    expect(resp).toHaveProperty("patch");
    expect(resp.testMode).toEqual(false);
  });

  it("publicKey", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****

    // Derivation path. First 3 items are automatically hardened!
    const path = [44, 931, 0, 0, 0];

    const resp = await app.getPublicKey(path);

    // console.log(resp);

    expect(resp.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(resp.errorMessage).toEqual("No errors");
    expect(resp).toHaveProperty("compressedPk");
    expect(resp.compressedPk.length).toEqual(33);
    expect(resp.compressedPk.toString("hex")).toEqual(THOR_PUB_KEY);
  });

  it("getAddressAndPubKey", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****

    // Derivation path. First 3 items are automatically hardened!
    const path = [44, 931, 0, 0, 0];
    const resp = await app.getAddressAndPubKey(path, "thor");

    // console.log(resp);

    expect(resp.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(resp.errorMessage).toEqual("No errors");

    expect(resp).toHaveProperty("bech32Address");
    expect(resp).toHaveProperty("compressedPk");

    expect(resp.bech32Address).toEqual(THOR_ADDRESS);
    expect(resp.compressedPk.length).toEqual(33);
  });

  it("appInfo", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger
    // *****
    const resp = await app.getAppInfo();

    // console.log(resp);

    expect(resp.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(resp.errorMessage).toEqual("No errors");

    expect(resp).toHaveProperty("appName");
    expect(resp).toHaveProperty("appVersion");
    expect(resp).toHaveProperty("flagLen");
    expect(resp).toHaveProperty("flagsValue");
    expect(resp).toHaveProperty("flagRecovery");
    expect(resp).toHaveProperty("flagSignedMcuCode");
    expect(resp).toHaveProperty("flagOnboarded");
    expect(resp).toHaveProperty("flagPinValidated");
  });

  it("deviceInfo", async () => {
    // *****
    // Requirements: Open `Dashboard` on Ledger (but not `THORChain` app)
    // *****
    const resp = await app.getDeviceInfo();

    console.log(resp);

    expect(resp.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(resp.errorMessage).toEqual("No errors");

    expect(resp).toHaveProperty("targetId");
    expect(resp).toHaveProperty("seVersion");
    expect(resp).toHaveProperty("flag");
    expect(resp).toHaveProperty("mcuVersion");
  });

  it("sign_and_verify_MsgSend", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger and follow instructions
    // *****

    // Derivation path. First 3 items are automatically hardened!
    const path = [44, 931, 0, 0, 0];
    const message = `{"account_number":"588","chain_id":"thorchain","fee":{"amount":[],"gas":"2000000"},"memo":"TestMemo","msgs":[{"type":"thorchain/MsgSend","value":{"amount":[{"amount":"150000000","denom":"rune"}],"from_address":"${THOR_ADDRESS}","to_address":"tthor10xgrknu44d83qr4s4uw56cqxg0hsev5e68lc9z"}}],"sequence":"5"}`;

    const responsePk = await app.getPublicKey(path);
    console.log("responsePk", responsePk);
    expect(responsePk.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(responsePk.errorMessage).toEqual("No errors");

    const responseSign = await app.sign(path, message);
    console.log("responseSign", responseSign);
    expect(responseSign.returnCode).toEqual(LedgerErrorType.NoErrors);
    expect(responseSign.errorMessage).toEqual("No errors");

    // // Check signature is valid
    const hash = crypto.createHash("sha256");
    const msgHash = hash.update(message).digest();

    const signatureDER = responseSign?.signature ?? Buffer.alloc(0);
    const signature = secp256k1.signatureImport(signatureDER);
    console.log(`signature`, signature);
    const signatureOk = secp256k1.ecdsaVerify(signature, msgHash, responsePk.compressedPk);
    // console.log('signatureOks', signatureOk);
    expect(signatureOk).toBeTruthy();
  });

  it("sign_invalid", async () => {
    // *****
    // Requirements: Open `THORChain` app on Ledger and follow instructions
    // *****

    const path = [44, 931, 0, 0, 0]; // Derivation path. First 3 items are automatically hardened!
    const invalidMessage = `{"chain_id":"thorchain","fee":{"amount":[],"gas":"10000000"},"memo":"","msgs":[{"type":"thorchain/MsgDeposit","value":{"coins":[{"amount":"330000000","asset":"THOR.RUNE"}],"memo":"SWAP:BNB.BNB:tbnb1qk2m905ypazwfau9cn0qnr4c4yxz63v9u9md20:","signer":"${THOR_ADDRESS}"}}],"sequence":"6"}`;

    const responseSign = await app.sign(path, invalidMessage);

    // console.log(responseSign);

    expect(responseSign.returnCode).toEqual(LedgerErrorType.DataIsInvalid);
    expect(responseSign.errorMessage).toEqual("Data is invalid : JSON Missing account number");
  });
});


tests/unit/basic.spec.ts
import { getBech32FromPK } from "../../src/index";

describe("Helpers unit", () => {
  it("check address conversion", async () => {
    const pkStr = "0264bff60f78aae3326b4fc3e16a7c48d0158bfe175fa4a9361190a42565662ae6";
    const pk = Buffer.from(pkStr, "hex");
    const addr = getBech32FromPK("thor", pk);
    expect(addr).toEqual("thor1mwyrp6lj85swy5e5g4hjlaacm33g6rw3p4qmq4");
  });
});


src/index.ts
/** ******************************************************************************
 *  (c) 2021 THORChain
 *  (c) 2019 ZondaX GmbH
 *  (c) 2016-2017 Ledger
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 ******************************************************************************* */

export * from "./types";
export * from "./helper";
export * from "./common";
import { THORChainApp } from "./app";
export default THORChainApp;


src/types.ts

export type BaseResponse = { returnCode: number; errorMessage: string };

export type VersionResponse = BaseResponse & {
  testMode: boolean;
  major: number;
  minor: number;
  patch: number;
  deviceLocked: boolean;
  targetId: string;
};

export type AppInfoResponse = BaseResponse &  {
  appName: string;
  appVersion: string;
  flagLen: number;
  flagsValue: number;
  flagRecovery: boolean;
  flagSignedMcuCode: boolean;
  flagOnboarded: boolean;
  flagPinValidated: boolean;
}

export type DeviceInfoResponse = BaseResponse &  {
  targetId: string
  seVersion: string
  flag: string
  mcuVersion: string
}

export type SignResponse = BaseResponse &  {
  signature?: Buffer
}

export type PubKeyResponse = BaseResponse &  {
  compressedPk: Buffer
}

export type AddressPubKeyResponse = PubKeyResponse &  {
  bech32Address: string
}


src/common.ts
export const CLA = 0x55;
export const CHUNK_SIZE = 250;
export const APP_KEY = "CSM";

export const INS = {
  GET_VERSION: 0x00,
  SIGN_SECP256K1: 0x02,
  GET_ADDR_SECP256K1: 0x04,
};

export const PAYLOAD_TYPE = {
  INIT: 0x00,
  ADD: 0x01,
  LAST: 0x02,
};

export const P1_VALUES = {
  ONLY_RETRIEVE: 0x00,
  SHOW_ADDRESS_IN_DEVICE: 0x01,
};

export class LedgerError extends Error {
  type: LedgerErrorType;
  constructor(type: LedgerErrorType, message: string = "") {
    super(`LedgerError ${type}: ${message ? message : errorCodeToString(type)}`);
    this.type = type;
  }

  toString() {
    return this.message;
  }
}

export enum LedgerErrorType {
  U2FUnknown = 1,
  U2FBadRequest = 2,
  U2FConfigurationUnsupported = 3,
  U2FDeviceIneligible = 4,
  U2FTimeout = 5,
  HPRInvalid = 10,
  PKInvalidBytes = 20,
  Timeout = 30,
  NoErrors = 0x9000,
  DeviceIsBusy = 0x9001,
  ErrorDerivingKeys = 0x6802,
  ExecutionError = 0x6400,
  WrongLength = 0x6700,
  EmptyBuffer = 0x6982,
  OutputBufferTooSmall = 0x6983,
  DataIsInvalid = 0x6984,
  ConditionsNotSatisfied = 0x6985,
  TransactionRejected = 0x6986,
  BadKeyHandle = 0x6a80,
  InvalidP1P2 = 0x6b00,
  InstructionNotSupported = 0x6d00,
  AppDoesNotSeemToBeOpen = 0x6e00,
  UnknownError = 0x6f00,
  SignVerifyError = 0x6f01,
  UnknownResponse = 0xffff,
}

export const LedgerErrorDescription: Record<LedgerErrorType, string> = {
  [LedgerErrorType.U2FUnknown]: "U2F: Unknown",
  [LedgerErrorType.U2FBadRequest]: "U2F: Bad request",
  [LedgerErrorType.U2FConfigurationUnsupported]: "U2F: Configuration unsupported",
  [LedgerErrorType.U2FDeviceIneligible]: "U2F: Device Ineligible",
  [LedgerErrorType.U2FTimeout]: "U2F: Timeout",
  [LedgerErrorType.Timeout]: "Timeout",
  [LedgerErrorType.HPRInvalid]: "Invalid HPR",
  [LedgerErrorType.PKInvalidBytes]: "Expected compressed public key [31 bytes]",
  [LedgerErrorType.NoErrors]: "No errors",
  [LedgerErrorType.DeviceIsBusy]: "Device is busy",
  [LedgerErrorType.ErrorDerivingKeys]: "Error deriving keys",
  [LedgerErrorType.ExecutionError]: "Execution Error",
  [LedgerErrorType.WrongLength]: "Wrong Length",
  [LedgerErrorType.EmptyBuffer]: "Empty Buffer",
  [LedgerErrorType.OutputBufferTooSmall]: "Output buffer too small",
  [LedgerErrorType.DataIsInvalid]: "Data is invalid",
  [LedgerErrorType.ConditionsNotSatisfied]: "Conditions not satisfied",
  [LedgerErrorType.TransactionRejected]: "Transaction rejected",
  [LedgerErrorType.BadKeyHandle]: "Bad key handle",
  [LedgerErrorType.InvalidP1P2]: "Invalid P1/P2",
  [LedgerErrorType.InstructionNotSupported]: "Instruction not supported",
  [LedgerErrorType.AppDoesNotSeemToBeOpen]: "App does not seem to be open",
  [LedgerErrorType.UnknownError]: "Unknown error",
  [LedgerErrorType.SignVerifyError]: "Sign/verify error",
  [LedgerErrorType.UnknownResponse]: "Unknown response",
};

const isLedgerErrorType = (code: number): code is LedgerErrorType => code in LedgerErrorDescription;

export function errorCodeToString(statusCode: LedgerErrorType): string {
  if (isLedgerErrorType(statusCode)) return LedgerErrorDescription[statusCode];
  return `Unknown Status Code: ${statusCode}`;
}

function isErrorResponse(v: unknown): v is { return_code: number; error_message: string } {
  return typeof v === "object" && !!v && "return_code" in v && "return_message" in v;
}

function isStatusCodeResponse(v: unknown): v is { statusCode: number } {
  return typeof v === "object" && !!v && "statusCode" in v;
}

export function ledgerErrorFromResponse(response?: unknown): LedgerError {
  if (isStatusCodeResponse(response)) {
    return new LedgerError(response.statusCode);
  }

  if (isErrorResponse(response)) {
    return new LedgerError(response.return_code, response.error_message);
  }

  return new LedgerError(LedgerErrorType.UnknownResponse, `Unknown response ${response}`);
}


src/helper.ts
import type Transport from "@ledgerhq/hw-transport";
import {
  CLA,
  errorCodeToString,
  INS,
  LedgerError,
  LedgerErrorType,
  PAYLOAD_TYPE,
  ledgerErrorFromResponse,
} from "./common";
import crypto from "crypto";
import Ripemd160 from "ripemd160";
import { bech32 } from "bech32";
import { PubKeyResponse, SignResponse, VersionResponse } from "./types";

export function serializeHRP(hrp: string): Buffer {
  if (!hrp || hrp.length < 3 || hrp.length > 83) {
    throw new LedgerError(LedgerErrorType.HPRInvalid);
  }
  const buf = Buffer.alloc(1 + hrp.length);
  buf.writeUInt8(hrp.length, 0);
  buf.write(hrp, 1);
  return buf;
}

export function getBech32FromPK(hrp: string, pk: Buffer): string {
  if (pk.length !== 33) {
    throw new LedgerError(LedgerErrorType.PKInvalidBytes);
  }

  const hashSha256 = crypto.createHash("sha256").update(pk).digest();
  const hashRip = new Ripemd160().update(hashSha256).digest();
  const words = bech32.toWords(hashRip);
  return bech32.encode(hrp, words);
}

export function serializePath(path: number[]): Buffer {
  if (!path || path.length !== 5) {
    throw new Error("Invalid path.");
  }

  const buf = Buffer.alloc(20);
  buf.writeUInt32LE(0x80000000 + path[0], 0);
  buf.writeUInt32LE(0x80000000 + path[1], 4);
  buf.writeUInt32LE(0x80000000 + path[2], 8);
  buf.writeUInt32LE(path[3], 12);
  buf.writeUInt32LE(path[4], 16);

  return buf;
}

export async function signSendChunk(
  transport: Transport,
  chunkIdx: number,
  chunkNum: number,
  chunk: Buffer,
): Promise<SignResponse> {
  let payloadType = PAYLOAD_TYPE.ADD;
  if (chunkIdx === 1) {
    payloadType = PAYLOAD_TYPE.INIT;
  }
  if (chunkIdx === chunkNum) {
    payloadType = PAYLOAD_TYPE.LAST;
  }

  const response: Buffer = await transport.send(CLA, INS.SIGN_SECP256K1, payloadType, 0, chunk, [
    LedgerErrorType.NoErrors,
    LedgerErrorType.DataIsInvalid,
    LedgerErrorType.BadKeyHandle,
    LedgerErrorType.SignVerifyError,
  ]);

  const errorCodeData = response.slice(-2);
  const returnCode = errorCodeData[0] * 256 + errorCodeData[1];
  let errorMessage = errorCodeToString(returnCode);

  if (
    returnCode === LedgerErrorType.BadKeyHandle ||
    returnCode === LedgerErrorType.DataIsInvalid ||
    returnCode === LedgerErrorType.SignVerifyError
  ) {
    errorMessage = `${errorMessage} : ${response.slice(0, response.length - 2).toString("ascii")}`;
    throw new LedgerError(returnCode, errorMessage);
  }

  if (returnCode === LedgerErrorType.NoErrors && response.length > 2) {
    return {
      returnCode,
      errorMessage,
      signature: response.slice(0, response.length - 2),
    };
  }

  return {
    returnCode,
    errorMessage,
  };
}

export async function getVersion(transport: Transport): Promise<VersionResponse> {
  try {
    const response: Buffer = await transport.send(CLA, INS.GET_VERSION, 0, 0);

    const errorCodeData = response.slice(-2);
    const returnCode = (errorCodeData[0] * 256 + errorCodeData[1]) as LedgerErrorType;

    let targetId = 0;
    if (response.length >= 9) {
      /* eslint-disable no-bitwise */
      targetId = (response[5] << 24) + (response[6] << 16) + (response[7] << 8) + (response[8] << 0);
      /* eslint-enable no-bitwise */
    }

    return {
      returnCode,
      errorMessage: errorCodeToString(returnCode),
      testMode: response[0] !== 0,
      major: response[1],
      minor: response[2],
      patch: response[3],
      deviceLocked: response[4] === 1,
      targetId: targetId.toString(16),
    };
  } catch (error) {
    throw ledgerErrorFromResponse(error);
  }
}

export async function getPublicKey(transport: Transport, data: Buffer): Promise<PubKeyResponse> {
  try {
    const response = await transport.send(CLA, INS.GET_ADDR_SECP256K1, 0, 0, data, [
      LedgerErrorType.NoErrors,
    ]);
    const errorCodeData = response.slice(-2);
    const returnCode = errorCodeData[0] * 256 + errorCodeData[1];
    const compressedPk = Buffer.from(response.slice(0, 33));

    return {
      compressedPk,
      returnCode,
      errorMessage: errorCodeToString(returnCode),
    };
  } catch (error) {
    throw ledgerErrorFromResponse(error);
  }
}

/**
 * Takes raw output from Ledger device which is TLV encoded in the format "0x30 L 0x02 Lr r 0x02 Ls s"
 * where L is length of entire message (after 2nd byte), Lr is length of r and Ls is length of s.
 * This function extracts r and s, ensures they are 32 bytes each, joins into a single 64 byte Array.
 * This value returned should be base64 encoded to a string to add to the JSON 'signatures' to send to RPC.
 *
 * Example raw sig:
 * [ 48, 69, 2, 33,
 *   0, 161, 207, 217, 127, 151, 190, 23, 98, 217, 186, 108, 82, 102, 19, 222, 20, 171, 6, 240, 134, 195, 251, 98, 190, 246, 228, 243, 215, 95, 166, 121, 165,
 *   2, 32,
 *   69, 188, 53,  213, 24, 9, 191, 90, 244, 21, 213, 146, 240, 109, 156, 221, 247, 63, 131, 52, 150, 253, 199, 153, 132, 76, 91, 239, 28, 254, 68, 80 ]
 *
 *   48 69  -- length is 69
 *   2 33   -- length of r is 33 bytes
 *   <33 bytes of r>  -- (the leading zero here can be dropped)
 *   2 32   -- length of s is 32 bytes
 *   <32 bytes of s>
 */
export function extractSignatureFromTLV(signatureArray: number[]): number[] {
  // Check Type Length Value encoding
  if (signatureArray.length < 64) {
    throw Error("Invalid Signature: Too short");
  }
  if (signatureArray[0] != 0x30) {
    throw Error("Invalid Ledger Signature TLV encoding: expected first byte 0x30");
  }
  if (signatureArray[1] + 2 != signatureArray.length) {
    throw Error("Invalid Signature: signature length does not match TLV");
  }
  if (signatureArray[2] != 0x02) {
    throw Error("Invalid Ledger Signature TLV encoding: expected length type 0x02");
  }

  // r signature
  const rLength = signatureArray[3];
  var rSignature = signatureArray.slice(4, rLength + 4);
  // Drop leading zero on some 'r' signatures that are 33 bytes.
  if (rSignature.length == 33 && rSignature[0] == 0) {
    rSignature = rSignature.slice(1, 33);
  } else if (rSignature.length == 33) {
    throw Error('Invalid signature: "r" too long');
  }
  // add leading zero's to pad to 32 bytes
  while (rSignature.length < 32) {
    rSignature.unshift(0);
  }

  // s signature
  if (signatureArray[rLength + 4] != 0x02) {
    throw Error("Invalid Ledger Signature TLV encoding: expected length type 0x02");
  }
  const sLength = signatureArray[rLength + 5];
  if (4 + rLength + 2 + sLength != signatureArray.length) {
    throw Error("Invalid Ledger Signature: TLV byte lengths do not match message length");
  }
  var sSignature = signatureArray.slice(rLength + 6, signatureArray.length);
  // Drop leading zero on 's' signatures that are 33 bytes. This shouldn't occur since ledger signs using "Small s" math. But just to be sure...
  if (sSignature.length == 33 && sSignature[0] == 0) {
    sSignature = sSignature.slice(1, 33);
  } else if (sSignature.length == 33) {
    throw Error('Invalid signature: "s" too long');
  }
  // add leading zero's to pad to 32 bytes
  while (sSignature.length < 32) {
    sSignature.unshift(0);
  }

  if (rSignature.length != 32 || sSignature.length != 32) {
    throw Error("Invalid signatures: must be 32 bytes each");
  }

  return rSignature.concat(sSignature);
}


src/app.ts
import type Transport from "@ledgerhq/hw-transport";
import * as Helpers from "./helper";
import {
  APP_KEY,
  CHUNK_SIZE,
  CLA,
  INS,
  LedgerError,
  errorCodeToString,
  P1_VALUES,
  LedgerErrorType,
  ledgerErrorFromResponse,
} from "./common";
import {
  AddressPubKeyResponse,
  AppInfoResponse,
  DeviceInfoResponse,
  PubKeyResponse,
  SignResponse,
  VersionResponse,
} from "./types";

export class THORChainApp {
  transport: Transport;

  constructor(transport: Transport, scrambleKey: string = APP_KEY) {
    if (!transport) {
      throw new Error("Transport has not been defined");
    }

    this.transport = transport;
    transport.decorateAppAPIMethods(
      this,
      ["getVersion", "sign", "getAddressAndPubKey", "appInfo", "deviceInfo", "getBech32FromPK"],
      scrambleKey,
    );
  }

  async serializePath(path: number[]): Promise<Buffer> {
    const version = await Helpers.getVersion(this.transport);

    if (version.returnCode !== LedgerErrorType.NoErrors) {
      throw new LedgerError(version.returnCode, version.errorMessage);
    }

    switch (version.major) {
      case 2:
        return Helpers.serializePath(path);
      default:
        throw new LedgerError(LedgerErrorType.ExecutionError, "App Version is not supported");
    }
  }

  async signGetChunks(path: number[], message: string): Promise<Buffer[]> {
    const serializedPath: Buffer = await this.serializePath(path);

    const chunks = [];
    chunks.push(serializedPath);
    const buffer = Buffer.from(message);

    for (let i = 0; i < buffer.length; i += CHUNK_SIZE) {
      let end = i + CHUNK_SIZE;
      if (i > buffer.length) {
        end = buffer.length;
      }
      chunks.push(buffer.slice(i, end));
    }

    return chunks;
  }

  async getVersion(): Promise<VersionResponse> {
    return Helpers.getVersion(this.transport);
  }

  async getAppInfo(): Promise<AppInfoResponse> {
    try {
      const response: Buffer = await this.transport.send(0xb0, 0x01, 0, 0);
      const errorCodeData = response.slice(-2);
      const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

      if (response[0] !== 1) {
        // Ledger responds with format ID 1. There is no spec for any format != 1
        throw new LedgerError(LedgerErrorType.DeviceIsBusy, "response format ID not recognized");
      }

      const appNameLen = response[1];
      const appName = response.slice(2, 2 + appNameLen).toString("ascii");
      let idx = 2 + appNameLen;
      const appVersionLen = response[idx];
      idx += 1;
      const appVersion = response.slice(idx, idx + appVersionLen).toString("ascii");
      idx += appVersionLen;
      const appFlagsLen = response[idx];
      idx += 1;
      const flagLen = appFlagsLen;
      const flagsValue = response[idx];

      return {
        returnCode,
        errorMessage: errorCodeToString(returnCode),
        appName,
        appVersion,
        flagLen,
        flagsValue,
        // eslint-disable-next-line no-bitwise
        flagRecovery: (flagsValue & 1) !== 0,
        // eslint-disable-next-line no-bitwise
        flagSignedMcuCode: (flagsValue & 2) !== 0,
        // eslint-disable-next-line no-bitwise
        flagOnboarded: (flagsValue & 4) !== 0,
        // eslint-disable-next-line no-bitwise
        flagPinValidated: (flagsValue & 128) !== 0,
      };
    } catch (error) {
      throw ledgerErrorFromResponse(error);
    }
  }

  async getDeviceInfo(): Promise<DeviceInfoResponse> {
    try {
      const response: Buffer = await this.transport.send(0xe0, 0x01, 0, 0, Buffer.from([]), [
        LedgerErrorType.NoErrors,
        LedgerErrorType.AppDoesNotSeemToBeOpen,
      ]);

      const errorCodeData = response.slice(-2);
      const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

      if (returnCode === LedgerErrorType.AppDoesNotSeemToBeOpen) {
        throw new LedgerError(returnCode, "This command is only available in the Dashboard");
      }

      const targetId = response.slice(0, 4).toString("hex");

      let pos = 4;
      const secureElementVersionLen = response[pos];
      pos += 1;
      const seVersion = response.slice(pos, pos + secureElementVersionLen).toString();
      pos += secureElementVersionLen;

      const flagsLen = response[pos];
      pos += 1;
      const flag = response.slice(pos, pos + flagsLen).toString("hex");
      pos += flagsLen;

      const mcuVersionLen = response[pos];
      pos += 1;
      // Patch issue in mcu version
      let tmp = response.slice(pos, pos + mcuVersionLen);
      if (tmp[mcuVersionLen - 1] === 0) {
        tmp = response.slice(pos, pos + mcuVersionLen - 1);
      }
      const mcuVersion = tmp.toString();

      return {
        returnCode,
        errorMessage: errorCodeToString(returnCode),
        targetId,
        seVersion,
        flag,
        mcuVersion,
      };
    } catch (error) {
      throw ledgerErrorFromResponse(error);
    }
  }

  async getPublicKey(path: number[]): Promise<PubKeyResponse> {
    try {
      const serializedPath = await this.serializePath(path);
      const version = await this.getVersion();

      switch (version.major) {
        case 2: {
          const data = Buffer.concat([Helpers.serializeHRP("thor"), serializedPath]);
          return Helpers.getPublicKey(this.transport, data);
        }
        default:
          throw new LedgerError(LedgerErrorType.ExecutionError, "App Version is not supported");
      }
    } catch (e) {
      throw ledgerErrorFromResponse(e);
    }
  }

  async getAddressAndPubKey(path: number[], hrp: string): Promise<AddressPubKeyResponse> {
    const serializedPath: Buffer = await this.serializePath(path);
    const serializedHRP = Helpers.serializeHRP(hrp);
    const data = Buffer.concat([serializedHRP, serializedPath]);
    const response: Buffer = await this.transport.send(
      CLA,
      INS.GET_ADDR_SECP256K1,
      P1_VALUES.ONLY_RETRIEVE,
      0,
      data,
      [LedgerErrorType.NoErrors],
    );
    const errorCodeData = response.slice(-2);
    const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

    const compressedPk = Buffer.from(response.slice(0, 33));
    const bech32Address = Buffer.from(response.slice(33, -2)).toString();

    return {
      bech32Address,
      compressedPk,
      returnCode,
      errorMessage: errorCodeToString(returnCode),
    };
  }

  async showAddressAndPubKey(path: number[], hrp: string): Promise<AddressPubKeyResponse> {
    const serializedPath: Buffer = await this.serializePath(path);
    const data = Buffer.concat([Helpers.serializeHRP(hrp), serializedPath]);
    const response: Buffer = await this.transport.send(
      CLA,
      INS.GET_ADDR_SECP256K1,
      P1_VALUES.SHOW_ADDRESS_IN_DEVICE,
      0,
      data,
      [LedgerErrorType.NoErrors],
    );
    const errorCodeData = response.slice(-2);
    const returnCode = errorCodeData[0] * 256 + errorCodeData[1];

    const compressedPk = Buffer.from(response.slice(0, 33));
    const bech32Address = Buffer.from(response.slice(33, -2)).toString();

    return {
      bech32Address,
      compressedPk,
      returnCode: returnCode,
      errorMessage: errorCodeToString(returnCode),
    };
  }

  async signSendChunk(chunkIdx: number, chunkNum: number, chunk: Buffer): Promise<SignResponse> {
    const version = await Helpers.getVersion(this.transport);
    switch (version.major) {
      case 2:
        return Helpers.signSendChunk(this.transport, chunkIdx, chunkNum, chunk);
      default:
        throw new LedgerError(LedgerErrorType.ExecutionError, "App Version is not supported");
    }
  }

  async sign(path: number[], message: string): Promise<SignResponse> {
    const chunks: Buffer[] = await this.signGetChunks(path, message);
    let response: SignResponse = await this.signSendChunk(1, chunks.length, chunks[0]);

    for (let i = 1; i < chunks.length; i += 1) {
      // eslint-disable-next-line no-await-in-loop
      response = await this.signSendChunk(1 + i, chunks.length, chunks[i]);
      if (response.returnCode !== LedgerErrorType.NoErrors) {
        break;
      }
    }

    return response;
  }
}
